tutorbackup/templates/backup/build/backup/restore_services.py

```diff
@@ -0,0 +1,189 @@
+    password = ENV['MYSQL_ROOT_PASSWORD']
+    host = ENV['MYSQL_HOST']
+    port = ENV['MYSQL_PORT']
+    dump_file = MYSQL_DUMPFILE
+
+    logger.info(f"Restoring MySQL databases on {host}:{port} from {dump_file}")
+    cmd = ("mysql "
+           f"--host={host} --port={port} "
+           f"--user={user} --password={password}")
+    with open(dump_file, 'rb') as dump:
+        check_call(cmd,
+                   shell=True,
+                   stdin=dump,
+                   stdout=sys.stdout,
+                   stderr=sys.stderr)
+
+    logger.info("MySQL restored.")
+
+
+def restore_mongodb():
+    host = ENV['MONGODB_HOST']
+    port = ENV['MONGODB_PORT']
+    dump_dir = MONGODB_DUMPDIR
+
+    logger.info(
+        f"Restoring MongoDB databases on {host}:{port} from {dump_dir}")
+    cmd = ("mongorestore "
+           "--stopOnError --drop "
+           f"--host={host} --port={port} "
+           f"{dump_dir}"
+           )
+    try:
+        cmd += (f" --username={ENV['MONGODB_USERNAME']} "
+                f"--password={ENV['MONGODB_PASSWORD']}")
+    except KeyError:
+        pass
+
+    check_call(cmd,
+               shell=True,
+               stdout=sys.stdout,
+               stderr=sys.stderr)
+
+    logger.info("MongoDB restored.")
+
+
+def restore_caddy():
+    dump_dir = CADDY_DUMPDIR
+    caddy_dir = 'caddy'
+    logger.info(f"Copying Caddy data from {dump_dir}")
+    shutil.copytree(dump_dir, caddy_dir)
+
+    total_size = get_size(caddy_dir)
+    logger.info(f"Complete. {caddy_dir} total size {total_size} bytes.")
+
+
+def extract():
+    tar_file = TARFILE
+    out_dir = DUMP_DIRECTORY
+
+    logger.info(f"Extracting archive {tar_file} to {out_dir}")
+    with tarfile.open(tar_file, "r:xz") as tar:
+        tar.extractall()
+
+    size = get_size(out_dir)
+    logger.info(f"Complete. {out_dir} is {size} bytes.")
+
+
+def download_from_s3(version_id=None):
+    from s3_client import S3_CLIENT, IntegrityError
+
+    bucket = ENV['S3_BUCKET_NAME']
+    file_name = TARFILE
+
+    logger.info(f"Downloading {file_name} from S3 bucket {bucket}")
+    try:
+        S3_CLIENT.download_file(
+            bucket,
+            os.path.basename(file_name),
+            file_name,
+            ExtraArgs={'VersionId': version_id} if version_id else None,
+        )
+
+        logger.info("Checking downloaded file's integrity ...")
+        obj_metadata = S3_CLIENT.head_object(
+            Bucket=bucket,
+            Key=os.path.basename(file_name),
+        )
+
+        received_size = obj_metadata['ContentLength']
+        size = os.path.getsize(file_name)
+        size_correct = received_size == size
+
+        received_version_id = obj_metadata['VersionId']
+        if version_id:
+            version_id_correct = version_id == received_version_id
+        else:
+            version_id_correct = True
+
+        received_checksum = obj_metadata['ETag'].strip('"')
+
+        if size_correct and version_id_correct:
```

Comment:
You can call me paranoid, but I'm not super fond of the idea of verifying by size alone. I also realize that integrity checking in S3 is a mess with `Content-MD5` vs. the various `x-amz-checksum-*` headers which bizarrely include several checksum algorithms *except* MD5 vs. [several other options](https://docs.aws.amazon.com/AmazonS3/latest/userguide/checking-object-integrity.html).

So, how about we just create a detached signature? As in, create `backup.tar.xz` and `backup.tar.xz.md5`?

Reply:
Could you elaborate on that? I don't know what a detached signature is.

Reply:
ah you mean that we upload both to S3 and check `backup.tar.xz.md5` upon downloading the backup? How does that work with versioning? Will both files have the same version ID?

Reply:
Ah, good point. No, they wont.

Reply:
Maybe we could add the checksum to object metadata as a key when uploading:
```
S3_CLIENT.upload_file(
            file_name,
            bucket,
            os.path.basename(file_name),
            extra_args={'Metadata': {'checksum': checksum_value}}
)
```
And then retrieve that when downloading and compare it to the downloaded file's checksum.

Reply:
I'll test this to see if it works

Reply:
You could calculate a SHA256 checksum, add that as `checksum-sha256` metadata, and then compare that against the `x-amz-checksum-256` header as an integrity check.

Reply:
> I'll test this to see if it works

You can also leave that for a separate PR if you would like me to merge this version as-is.

Reply:
> You could calculate a SHA256 checksum, add that as `checksum-sha256` metadata, and then compare that against the `x-amz-checksum-256` header as an integrity check.

I have tried that before. I didn't get a checksum back. There are also numerous other ways in boto3 but they are either not documented properly or I wasn't able to get them to work.

Reply:
> Maybe we could add the checksum to object metadata as a key when uploading:
> 
> ```
> S3_CLIENT.upload_file(
>             file_name,
>             bucket,
>             os.path.basename(file_name),
>             extra_args={'Metadata': {'checksum': checksum_value}}
> )
> ```
> 
> And then retrieve that when downloading and compare it to the downloaded file's checksum.

It works. I think this is the best solution.

Reply:
Excellent. In that case my only request would be to call the metadata field `checksum-<algo>` as opposed to just `checksum`. So for example if you do an MD5 hash, make it `checksum-md5`.

Reply:
will do